File: taller_hilos/ejercicio8/Ejercicio8.java
// Ejercicio 8: Hilos con pausas usando sleep
//
// Ilustra cómo distintas velocidades de sleep producen ritmos
// de ejecución diferentes en hilos concurrentes.
// El hilo rápido terminará antes aunque hayan arrancado al mismo tiempo.

class TareaConPausa implements Runnable {
    private String nombre;
    private int pausaMs; // Tiempo de pausa en milisegundos entre pasos

    TareaConPausa(String nombre, int pausaMs) {
        this.nombre = nombre;
        this.pausaMs = pausaMs;
    }

    public void run() {
        for (int i = 1; i <= 5; i++) {
            System.out.println(nombre + " - paso " + i);
            try {
                // Cada hilo usa su propio tiempo de pausa.
                // sleep() no bloquea a los otros hilos, solo a este.
                Thread.sleep(pausaMs);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        System.out.println(nombre + " finalizado.");
    }
}

public class Ejercicio8 {
    public static void main(String[] args) {
        // Hilo lento: 1 segundo entre pasos → tarda ~5 s en total.
        Thread lento  = new Thread(new TareaConPausa("Hilo-Lento",  1000));
        // Hilo rápido: 300 ms entre pasos → tarda ~1.5 s en total.
        Thread rapido = new Thread(new TareaConPausa("Hilo-Rapido",  300));

        // Ambos arrancan a la vez; el rápido terminará primero.
        lento.start();
        rapido.start();
    }
}

File: taller_hilos/ejercicio8/ejercicio8.py
import threading
import time

# Ejercicio 8: Hilos con pausas usando sleep
#
# Dos hilos realizan la misma tarea pero con tiempos de pausa distintos.
# Observar la salida muestra claramente que corren en paralelo y que
# el hilo rápido adelanta pasos mientras el lento aún está en los primeros.

def tarea_con_pausa(nombre, pausa):
    for i in range(1, 6):
        print(f"{nombre} - paso {i}")
        # time.sleep() suspende solo este hilo durante 'pausa' segundos.
        time.sleep(pausa)
    print(f"{nombre} finalizado.")

# Hilo lento: 1 s de pausa por paso → ~5 s en total.
lento  = threading.Thread(target=tarea_con_pausa, args=("Hilo-Lento",  1.0))
# Hilo rápido: 0.3 s de pausa por paso → ~1.5 s en total.
rapido = threading.Thread(target=tarea_con_pausa, args=("Hilo-Rapido", 0.3))

lento.start()
rapido.start()

# Esperamos a ambos. El rápido terminará y join() regresará enseguida;
# luego esperamos al lento el tiempo que le quede.
lento.join()
rapido.join()
